pageserver/src/tenant.rs

```diff
@@ -1350,13 +1268,64 @@ impl Tenant {
                         .map_err(LoadLocalTimelineError::ResumeDeletion);
                     }
 
-                    // We're loading fresh timeline that didnt yet make it into remote.
-                    (None, Some(remote_client))
+                    // see if there is at most one layer file and a metadata file in the directory
+                    // if not, we should fail the loading
+                    //
+                    // otherwise delete the not yet uploaded timeline.
+                    //
+                    // this timeline cannot have children because those are not allowed to be
+                    // created before uploading.
+
+                    let path = self.conf.timeline_path(&self.tenant_id, &timeline_id);
+
+                    return tokio::task::spawn_blocking({
+                        let path = path.clone();
+                        move || {
+                            let mut metadata = false;
+                            let mut layers = 0;
+                            let mut others = 0;
+                            for dentry in path.read_dir_utf8()? {
+                                let dentry = dentry?;
+                                let file_name = dentry.file_name();
+
+                                if file_name == "metadata" {
+                                    metadata = true;
+                                    continue;
+                                }
+
+                                use std::str::FromStr;
+
+                                if crate::tenant::storage_layer::LayerFileName::from_str(file_name)
+                                    .is_ok()
+                                {
+                                    layers += 1;
+                                    continue;
+                                }
+
+                                others += 1;
+                            }
+
+                            // bootstrapped have the one image layer file, or one partial temp file
+                            if !(metadata && layers + others <= 1) {
+                                anyhow::bail!("unexpected assumed fresh timeline: found metadata={}, layers={}, others={}", metadata, layers, others);
+                            }
```

Comment:
Is this even needed? I was wondering about the worst case, we could remove something which was not supposed to be there perhaps with the transitional period, but I think the #5484 was already released, so this shouldn't be an issue, right?

If we do think that there's a real risk, then this should probably be collecting all of those files, or at least the "others".

Rationale here is that a bootstrapped timeline will have either:
- ready image layer file
- partial not yet renamed temp file
- metadata

An branched timeline will have:
- metadata

If we crash earlier the timeline will be uninit, and we've already deleted it in `Timeline::scan_...`.

Reply:
Because the timeline creation API waits for remote index upload, we should never hit this `bail` unless the control plane misbehaves and allows some endpoints to exist before the timeline creation request has returned success.

So I'd suggest we keep this check, and think of it as a sanity check for bugs elsewhere in the system, rather than as something we would expect to see in the field.